neqo-udp/src/lib.rs

```diff
@@ -8,37 +8,133 @@
 #![allow(clippy::missing_panics_doc)] // Functions simply delegate to tokio and quinn-udp.
 
 use std::{
+    cell::RefCell,
     io::{self, IoSliceMut},
-    net::{SocketAddr, ToSocketAddrs},
+    net::SocketAddr,
     slice,
 };
 
-use neqo_common::{Datagram, IpTos};
+use neqo_common::{qdebug, Datagram, IpTos};
 use quinn_udp::{EcnCodepoint, RecvMeta, Transmit, UdpSocketState};
-use tokio::io::Interest;
 
 /// Socket receive buffer size.
 ///
 /// Allows reading multiple datagrams in a single [`Socket::recv`] call.
 const RECV_BUF_SIZE: usize = u16::MAX as usize;
 
-pub struct Socket {
+std::thread_local! {
+    static RECEIVE_BUFFER: RefCell<Vec<u8>> = RefCell::new(vec![0; RECV_BUF_SIZE]);
+}
+
+pub struct Socket<S> {
+    state: UdpSocketState,
     #[allow(unknown_lints)] // available with Rust v1.75
     #[allow(clippy::struct_field_names)]
-    socket: tokio::net::UdpSocket,
-    state: UdpSocketState,
-    recv_buf: Vec<u8>,
+    socket: S,
+}
+
+impl<S> Socket<S> {
+    fn send_inner(
+        state: &UdpSocketState,
+        socket: quinn_udp::UdpSockRef<'_>,
+        d: &Datagram,
+    ) -> io::Result<()> {
+        let transmit = Transmit {
+            destination: d.destination(),
+            ecn: EcnCodepoint::from_bits(Into::<u8>::into(d.tos())),
+            contents: d,
+            segment_size: None,
+            src_ip: None,
+        };
+
+        state.send(socket, &transmit)?;
+
+        qdebug!(
+            "sent {} bytes from {} to {}",
+            d.len(),
+            d.source(),
+            d.destination()
+        );
+
+        Ok(())
+    }
+
+    fn recv_inner(
+        local_address: &SocketAddr,
+        state: &UdpSocketState,
+        socket: quinn_udp::UdpSockRef<'_>,
+    ) -> Result<Vec<Datagram>, io::Error> {
+        let mut meta = RecvMeta::default();
```

Comment:
This can move inside the closure.

Reply:
:+1: Done in https://github.com/mozilla/neqo/pull/1920/commits/af648706268ffee40e0206d6358828a511cfe4d3.